### ISSUE TITLE
Build error of AFALG engine on linux-armv4

### ISSUE THREAD

**Issue by shigeki** [openssl/openssl#890]
I had a build error in linking AFALG engine on linux-arm4 of raspberry Pi2 as
```sh
./libcrypto.a(init.o): In function `ossl_init_engine_afalg':
init.c:(.text+0x20): undefined reference to `engine_load_afalg_internal'
collect2: error: ld returned 1 exit status
Makefile.shared:186: recipe for target 'link_app.' failed
make[1]: *** [link_app.] Error 1
````
This is because compiling e_afalg.c was skipped with the warning of
```sh
gcc -DDSO_DLFCN -DHAVE_DLFCN_H -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines\"" -Wall -O3 -pthread  -march=armv7-a -Wa,--noexecstack -fPIC -Iinclude -I. -Icrypto/include -MMD -MF engines/afalg/e_afalg.d.tmp -MT engines/afalg/e_afalg.o -c -o engines/afalg/e_afalg.o engines/afalg/e_afalg.c
engines/afalg/e_afalg.c:69:3: warning: #warning "AFALG ENGINE requires Kernel Headers >= 4.1.0" [-Wcpp]
 # warning "AFALG ENGINE requires Kernel Headers >= 4.1.0"
   ^
engines/afalg/e_afalg.c:70:3: warning: #warning "Skipping Compilation of AFALG engine" [-Wcpp]
 # warning "Skipping Compilation of AFALG engine"
   ^
````
There is an inconsistency of kernel version between in process and in header file. The header file is still kernel version 3.10.
```sh
ohtsu@raspberrypi:~/github/openssl$ uname -r
4.1.17-v7+
ohtsu@raspberrypi:~/github/openssl$ cat /usr/include/linux/version.h
#define LINUX_VERSION_CODE 200711
#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))
ohtsu@raspberrypi:~/github/openssl$ apt-cache showpkg linux-libc-dev
Package: linux-libc-dev
Versions:
3.16.7-ckt11-1+deb8u4+rpi1 (/var/lib/apt/lists/mirrordirector.raspbian.org_raspbian_dists_jessie_main_binary-armhf_Packages) (/var/lib/dpkg/status)
````
The build was finished sucessful if I commented out kernel version check in  e_afalg.c, but the afalg test was failed with the following messages. 
```sh
../test/recipes/30-test_afalg.t ...........
Message from syslogd@raspberrypi at Mar 17 14:09:44 ...
 kernel:[  262.525420] Internal error: Oops - BUG: 0 [#1] SMP ARM

Message from syslogd@raspberrypi at Mar 17 14:09:44 ...
 kernel:[  262.614969] Process afalgtest (pid: 3628, stack limit = 0xb6744210)

Message from syslogd@raspberrypi at Mar 17 14:09:44 ...
 kernel:[  262.621221] Stack: (0xb6745db0 to 0xb6746000)
````

**Comment by jiapei100** [comment]
Great topic... I met exactly the same issue as yours, on Ubuntu 14.04.4 ...
Check my issue https://github.com/openssl/openssl/issues/912

`./libcrypto.a(init.o): In function `ossl_init_engine_afalg':
init.c:(.text+0x31): undefined reference to `engine_load_afalg_internal'
collect2: error: ld returned 1 exit status
make[1]: *** [link_app.] Error 1
make[1]: Leaving directory `/home/jiapei/Downloads/networks/openssl'
make: *** [apps/openssl] Error 2
`


```
jiapei@jiapei-GT72-6QE:~/Downloads/networks/openssl$ uname -r
4.2.0-34-generic
jiapei@jiapei-GT72-6QE:~/Downloads/networks/openssl$ cat /usr/include/linux/version.h
#define LINUX_VERSION_CODE 199947
#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))
jiapei@jiapei-GT72-6QE:~/Downloads/networks/openssl$ apt-cache showpkg linux-libc-dev
Package: linux-libc-dev
Versions: 
3.13.0-83.127 (/var/lib/apt/lists/ca.archive.ubuntu.com_ubuntu_dists_trusty-updates_main_binary-amd64_Packages) (/var/lib/apt/lists/security.ubuntu.com_ubuntu_dists_trusty-security_main_binary-amd64_Packages) (/var/lib/dpkg/status)
...
```

We may have to check ubuntuforums instead of here???


Cheers
Pei

**Comment by mattcaswell** [comment]
As a temporary workaround until this gets fixed you can disable afalg by passing the no-afalgeng parameter to Configure.

**Comment by shigeki** [comment]
Yes.
```sh
$./config no-afalgeng
````
works fine.

**Comment by jiapei100** [comment]
Yes, Great !! This does the trick.. 
Thank you ...

**Comment by jhcastel** [comment]
I am installing on a Raspberry PI 3, after option 
    $./config no-afalgeng
I still have the same error

**Comment by mattcaswell** [comment]
@jhcastel, please can you be more specific? Which of the above errors do you see?

**Comment by jhcastel** [comment]
After running with option
`$./config no-afalgeng`
I got the following error 

`gcc -DDSO_DLFCN -DHAVE_DLFCN_H -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="/usr/local/ssl" -DENGINESDIR="/usr/local/lib/engines" -Wall -O3 -pthread -march=armv7-a -Wa,--noexecstack -o apps/openssl apps/app_rand.o apps/apps.o apps/asn1pars.o apps/ca.o apps/ciphers.o apps/cms.o apps/crl.o apps/crl2p7.o apps/dgst.o apps/dhparam.o apps/dsa.o apps/dsaparam.o apps/ec.o apps/ecparam.o apps/enc.o apps/engine.o apps/errstr.o apps/gendsa.o apps/genpkey.o apps/genrsa.o apps/nseq.o apps/ocsp.o apps/openssl.o apps/opt.o apps/passwd.o apps/pkcs12.o apps/pkcs7.o apps/pkcs8.o apps/pkey.o apps/pkeyparam.o apps/pkeyutl.o apps/prime.o apps/rand.o apps/rehash.o apps/req.o apps/rsa.o apps/rsautl.o apps/s_cb.o apps/s_client.o apps/s_server.o apps/s_socket.o apps/s_time.o apps/sess_id.o apps/smime.o apps/speed.o apps/spkac.o apps/srp.o apps/ts.o apps/verify.o apps/version.o apps/x509.o -L. -lssl -L. -lcrypto -ldl
./libcrypto.a(init.o): In function `ossl_init_engine_afalg':
init.c:(.text+0x20): undefined reference to `engine_load_afalg_internal'
collect2: error: ld returned 1 exit status`

**Comment by mattcaswell** [comment]
@jhcastel have you tried that against the latest master version? Also can you check the contents of the file include/openssl/opensslconf.h after running the above config command. Can you confirm that you see the following block in there:
````
#ifndef OPENSSL_NO_AFALGENG
# define OPENSSL_NO_AFALGENG
#endif
````

**Comment by jhcastel** [comment]
that block is in the file include/openssl/opensslconf.h, but it did not work.
I deleted the whole folder and clone again from the master. Run again the configuration script with no-afalgeng option And at the end it works.

Thanks a lot

**Comment by mattcaswell** [comment]
This issue should be fixed by 627537ddf379.

**Comment by shigeki** [comment]
I've found that the build error is resolved and make tests are fine. Thanks.
